gnomad/utils/constraint.py

```diff
@@ -563,6 +568,247 @@ def collapse_strand(
+        values as corresponding allele number Tables. Default is None.
+    :param transformation_funcs: An optional Dictionary to transform the HTs before
+        adding annotations. Default is None, which will use some default transformations
+        as described above.
+    :return: Table with sites split and necessary annotations.
+    """
+    ref = get_reference_genome(ht.locus)
+    if ref == hl.get_reference("GRCh37"):
+        from gnomad.resources.grch37 import transform_grch37_methylation as trans_methyl
+    else:
+        from gnomad.resources.grch38 import transform_grch38_methylation as trans_methyl
+
+    # Check if context Table is split, and if not, split multiallelic sites.
+    if "was_split" not in list(ht.row):
+        ht = hl.split_multi_hts(ht)
+
+    # Only need to keep the keys (locus, alleles) and the context and vep annotations.
+    ht = ht.select("context", "vep")
+
+    # Filter Table to only contigs 1-22, X, Y.
+    ht = hl.filter_intervals(
+        ht, [hl.parse_locus_interval(c, ref.name) for c in ref.contigs[:24]]
+    )
+
+    # Add annotations for 'ref' and 'alt'.
+    ht = ht.annotate(ref=ht.alleles[0], alt=ht.alleles[1])
+
+    # Trim heptamer context to create trimer context and filter to where the bases are
+    # either A, T, C, or G.
+    ht = trimer_from_heptamer(ht)
+    ht = ht.filter(ht.context.matches(f"[ATCG]{{{3}}}"))
+
+    # Annotate mutation type (such as "CpG", "non-CpG transition", "transversion") and
+    # collapse strands to deduplicate the context.
+    ht = annotate_mutation_type(collapse_strand(ht))
+
+    # Add most_severe_consequence annotation to 'transcript_consequences' within the
+    # vep root annotation.
+    csqs = ht.vep.transcript_consequences
+    csqs = add_most_severe_consequence_to_consequence(csqs)
+    vep_csq_fields = [
+        "transcript_id",
+        "gene_id",
+        "gene_symbol",
+        "biotype",
+        "most_severe_consequence",
+        "mane_select",
+        "canonical",
+        "lof",
+        "lof_flags",
+    ]
+    vep_csq_fields = [x for x in vep_csq_fields if x in csqs.dtype.element_type]
+    ht = ht.annotate(
+        vep=ht.vep.select(
+            "most_severe_consequence",
+            transcript_consequences=csqs.map(lambda x: x.select(*vep_csq_fields)),
+        )
+    )
+
+    # Add 'methylation_level', 'coverage', and 'gerp' annotations if specified.
+    transformation_funcs = transformation_funcs or {}
+
+    if "methylation_level" not in transformation_funcs:
+        transformation_funcs["methylation_level"] = lambda x, t: hl.if_else(
+            ht.cpg, trans_methyl(x)[t.locus].methylation_level, 0
+        )
+
+    if "gerp" not in transformation_funcs:
+        transformation_funcs["gerp"] = lambda x, t: hl.if_else(
+            hl.is_missing(x[t.locus].S), 0, x[t.locus].S
+        )
+
+    # If necessary, pull out first element of coverage statistics (which includes all
```

Comment:
The comment specifies that this is relevant to v4...

It's added here: https://github.com/broadinstitute/gnomad_methods/blob/d29acb350ee202b287d9c7633f3e0ccf62d061e5/gnomad/utils/sparse_mt.py#L1334

Reply:
I meant when I tried to get `coverage('exomes').ht()` for v4, I don't see `coverage_stats` as you mentioned in the comment, but the nested rows are there. Are there other versions of coverage table?

Reply:
I see there's a difference between GRCh37 and GRCh38 coverage table: the columns changed from: 
```
Row fields:
    'row_id': int64 
    'locus': locus<GRCh37> 
    'mean': float64 
    'median': int32 
```
to: 
```
Row fields:
    'locus': locus<GRCh38> 
    'mean': float64 
    'median_approx': int32 
    'total_DP': int64 
```

Reply:
It's in this one: "gs://gnomad/release/4.0/ht/exomes/gnomad.exomes.v4.0.coverage.ht"

```
from gnomad_qc.v4.resources.release import release_coverage
ht = release_coverage("exomes", public=False).ht()
ht.describe()
```

```
----------------------------------------
Global fields:
    'coverage_stats_meta': array<dict<str, str>> 
    'coverage_stats_meta_sample_count': array<int32> 
----------------------------------------
Row fields:
    'locus': locus<GRCh38> 
    'coverage_stats': array<struct {
        mean: float64, 
        median_approx: int32, 
        total_DP: int64, 
        over_1: float64, 
        over_5: float64, 
        over_10: float64, 
        over_15: float64, 
        over_20: float64, 
        over_25: float64, 
        over_30: float64, 
        over_50: float64, 
        over_100: float64
    }> 
----------------------------------------
Key: ['locus']
----------------------------------------
```

The point is that it can work with either possible schema

Reply:
Okay, that's what I want to know, thanks.